pkg/sql/sem/tree/uri.go

```diff
@@ -0,0 +1,128 @@
+// Copyright 2024 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package tree
+
+import (
+	"github.com/cockroachdb/cockroach/pkg/cloud/uris"
+	"github.com/cockroachdb/errors"
+)
+
+// URI represents a string expression containing a URL. During formatting, the
+// URL is sanitized to remove sensitive information such as passwords and keys.
+type URI struct {
+	Expr      Expr
+	kms       bool
+	sanitized bool
+}
+
+type URIs []URI
+
+var _ NodeFormatter = URI{}
+var _ NodeFormatter = URIs{}
+
+// NewURI constructs a URI wrapper around the expression, which is a string
+// literal containing an unsantized URL.
+func NewURI(e Expr) URI {
+	return URI{Expr: e}
+}
+
+// NewKMSURI constructs a URI wrapper around the expression, which is a string
+// literal containing an unsantized KMS URL.
+func NewKMSURI(e Expr) URI {
+	return URI{Expr: e, kms: true}
+}
+
+// NewSanitizedURI constructs a URI wrapper around the string, which is a
+// sanitized URL that should not be re-sanitized.
+func NewSanitizedURI(sanitized string) URI {
+	return URI{Expr: NewStrVal(sanitized), sanitized: true}
+}
+
+// NewSanitizedKMSURI constructs a URI wrapper around the string, which is a
+// sanitized KMS URL that should not be re-sanitized.
+func NewSanitizedKMSURI(sanitized string) URI {
+	return URI{Expr: NewStrVal(sanitized), kms: true, sanitized: true}
+}
+
+// Format implements the NodeFormatter interface. If the URI wraps a string
+// literal, we sanitize the URL contained in the string literal.
+func (uri URI) Format(ctx *FmtCtx) {
+	switch n := uri.Expr.(type) {
+	case *StrVal, *DString:
+		if uri.sanitized ||
+			ctx.HasAnyFlags(FmtShowPasswords|FmtHideConstants|FmtConstantsAsUnderscores) {
+			ctx.FormatNode(n)
+			return
+		}
+		var raw string
+		var sanitized string
+		var err error
+		if str, ok := uri.Expr.(*StrVal); ok {
+			raw = str.RawString()
+		} else {
+			raw = string(MustBeDString(uri.Expr))
+		}
+		if raw == "" || raw == "_" {
+			// Some commands treat empty URIs as special. And if we've re-parsed a URI
+			// formatted with FmtHideConstants, we should not try to interpret it as a
+			// URL but should leave it as-is.
+			ctx.FormatNode(n)
+			return
+		}
+		if uri.kms {
+			sanitized, err = uris.RedactKMSURI(raw)
+		} else {
+			sanitized, err = uris.SanitizeExternalStorageURI(raw, nil)
+		}
+		if err != nil {
```

Comment:
There hasn't been a thorough discussion yet, so we can have it now.

The sanitization performed by `SanitizeExternalStorageURI` is already used by [backup](https://github.com/cockroachdb/cockroach/blob/d41cb10e932e9ecf9b7ac78336a6f81651ff3f76/pkg/ccl/backupccl/backup_planning.go#L159), [changefeeds](https://github.com/cockroachdb/cockroach/blob/d41cb10e932e9ecf9b7ac78336a6f81651ff3f76/pkg/ccl/changefeedccl/changefeed.go#L111), [create external connection](https://github.com/cockroachdb/cockroach/blob/d41cb10e932e9ecf9b7ac78336a6f81651ff3f76/pkg/sql/create_external_connection.go#L167), and [import](https://github.com/cockroachdb/cockroach/blob/d41cb10e932e9ecf9b7ac78336a6f81651ff3f76/pkg/sql/importer/import_planning.go#L238), so it seemed natural to me to also use it here. But it did require the first two commits (breaking circular dependencies). This seemed reasonable to me, since the compiler could help me perform the first two commits mechanically and confirm their correctness, but I understand that it creates a lot of code churn.

Pros:
- logged sanitized URL matches sanitized URL in SHOW output, job details, error messages, other places
- logged sanitized URL keeps non-sensitive information intact (domain, path, etc)
- same URL sanitization routine used everywhere, can just change that one routine

Cons:
- first two commits create code churn
- backport more difficult, though I would argue not necessarily more risky (if it compiles after the first two commits, they should be correct)

WDYT?

Reply:
The granular per-param sanitization is performed in backup and friends since we thought it was necessary -- when showing you jobs, we think we need to show you where the job is backing up to, or what options storage options it is using when those are not sensitive. But in a generic log of all queries/slow queries/etc, I wonder if that still holds, or if all we really need to indicate is that it is a backup statement (and you can go consult the jobs if you want to see the details of the backup)?  

> same URL sanitization routine used everywhere, can just change that one routine

FWIW, the alternative I'm considering still only has one routine for _sanitizing_ a URI; I'd just be eliding it completely if I don't use that routine rather than having a distinct routine.

I'm not strongly opposed to the first two commits; just trying to think about the "change as little code as possible" backport guideline and if we really need sophisticated granular sanitization of URIs here at all, or could just elide them.

Reply:
> I'm not strongly opposed to the first two commits; just trying to think about the "change as little code as possible" backport guideline and if we really need sophisticated granular sanitization of URIs here at all, or could just elide them.

Good point. @rytaft and @mgartner have also asked if I can figure out a simpler version for backport that elides the URLs completely. I will try that out, and then rebase this work on top of that simpler backportable change.